lib/Backends/Interpreter/InterpreterFunction.cpp

```diff
@@ -229,21 +231,55 @@ Error BoundInterpreterFunction::execute(IRFunction *F,
     }
   }
 
-// Do the forward pass.
+  // Do the forward pass.
+  if (!getIRInstructionProcessingHandler()) {
+    // This is the fast version without hooks.
+    // Dispatch the interpreter on each instruction in the program:
+    for (const auto &I : F->getInstrs()) {
+      switch (I.getKind()) {
 #define DEF_VALUE(CLASS, NAME)
 #define DEF_INSTR(CLASS, NAME)                                                 \
   case Kinded::Kind::CLASS##Kind: {                                            \
     fwd##CLASS(llvm::cast<CLASS>(&I));                                         \
     break;                                                                     \
   }
 #define DEF_BACKEND_SPECIFIC_INSTR(CLASS, NAME)
-  // Dispatch the interpreter on each instruction in the program:
-  for (const auto &I : F->getInstrs()) {
-    switch (I.getKind()) {
 #include "glow/AutoGenInstr.def"
 
-    default:
-      llvm_unreachable("Invalid instruction.");
+      default:
+        glow::errs() << "Invalid instruction: " << &I << "\n";
+        llvm_unreachable("Invalid instruction.");
+      }
+    }
+  } else {
+    auto &irInstructionProcessingHandler = getIRInstructionProcessingHandler();
+    // Dispatch the interpreter on each instruction in the program:
+    for (const auto &I : F->getInstrs()) {
+      // Perform pre-processing.
+      irInstructionProcessingHandler(
+          &I, IRInstructionProcessingStage::PREPROCESSING, this);
+      // Perform processing and proceed with standard processing if needed.
+      if (!irInstructionProcessingHandler(
+              &I, IRInstructionProcessingStage::PROCESSING, this)) {
+        switch (I.getKind()) {
+#define DEF_VALUE(CLASS, NAME)
+#define DEF_INSTR(CLASS, NAME)                                                 \
+  case Kinded::Kind::CLASS##Kind: {                                            \
+    fwd##CLASS(llvm::cast<CLASS>(&I));                                         \
+    break;                                                                     \
+  }
+#define DEF_BACKEND_SPECIFIC_INSTR(CLASS, NAME)
+#include "glow/AutoGenInstr.def"
+
+        default:
+          glow::errs() << "Invalid instruction: " << &I << "\n";
+          llvm_unreachable("Invalid instruction.");
+        }
```

Comment:
nit: This logic is shared with the version without hooks. Could outline to a helper.